### ISSUE TITLE
PKCS11 support for user authentication

### ISSUE THREAD

**Issue by jfrederickson** [mumble-voip/mumble#4044]
**Context**
Certificates represent a user's identity in Mumble, and the consequences of a key compromise in this case are severe. If using Mumble on a machine you do not fully trust, it would be nice to have the ability to use e.g. a Yubikey to authenticate to Mumble servers without exposing the private key to the machine you're using. (This would also make it fairly easy to carry a Mumble identity with you between machines if you already regularly use a Yubikey, as I do.)

**Describe the feature you have in mind**
If a PKCS11 smartcard with an installed certificate is inserted, Mumble would allow users to use that for authentication rather than a cert/key on disk.

**Describe alternatives you've considered**
FIDO2 is a possible alternative, and FIDO2 keys are typically cheaper than PKCS11 smartcards. However, Mumble's auth system is already heavily reliant on X509 client certs; PKCS11 seems like it fits in better with the way Mumble works right now.

**Comment by Krzmbrzl** [comment]
I agree that this does sound useful to have and it might not even be that hard to implement (don't know that though).
However I have the feeling that we won't have time to implement this any time soon.

Thus if you or someone else that might stumble upon this issue want to give it a shot, we'd be glad to accept and help with a PullRequest for this :)

**Comment by davidebeatrici** [comment]
Right now Mumble saves the certificate in the user's "documents" folder after generating it.

We should ask the user whether he wants to do that and allow to specify a different path instead.

**Comment by stman** [comment]
Any news for this feature ? This would be really great to have it, because it would be a strong way to prevent MITM. AI voice based MITM attacks are terrible. This is a simple way to offer a first good level of protection against identity theft with mumble. Writting certificates on disks is a suicide so much our PC's, BIOS/UEFI/OS are pretty unsecure and full of backdoors.

**Comment by Krzmbrzl** [comment]
No news, no. There are also no plans to implement this right now. Thus, this feature is likely dependent on an external contributor working on it.

**Comment by zhiyan114** [comment]
It seems like I hit a roadblock while implementing this with Windows Certificate Store (see [QTBUG-131483](https://bugreports.qt.io/browse/QTBUG-131483)). For exportable private key, pulling the private key blobs should work. But unfortunately, we're working with non-exportable, so I'll have to wait until that resolves.

**Comment by zhiyan114** [comment]
Take that back, using schannel as socket backend fixes this issue.

@Krzmbrzl For PKCS11 directly, I'll probably need more time to figure things out. For windows systems, using the certificate store allows both smart card and existing certificates in there.

Is this implementation acceptable ([repo](https://github.com/zhiyan114/mumble/tree/sketch2))? There's definitely some cleanups that can be done after everything else is done.

https://github.com/user-attachments/assets/4e03e9e5-f88a-49fc-b2c3-df199e77e77d
